enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/ExecutionContextInstanceCache.scala

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compatibility.v3_4.runtime
+
+import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
+
+trait ExecutionContextInstanceCache[T, C <: ExecutionContext] {
+  self: T =>
+
+  type ExecutionContextConstructor = T => C
+
+  private var freeExecutionContextInstance: C = null.asInstanceOf[C]
```

Comment:
Having vars in a trait looks dodgy to me, can't solve this nicer?

Reply:
This trait is deleted now, but the point was for the using class to not have to be bothered with the implementation. Would you have been more ok to call it an abstract class? Wouldn't that just unnecessarily limit the mix-in possibilities?